src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs

```diff
@@ -0,0 +1,487 @@
+            if (table != null)
+            {
+                int index = KeyToBucket(table, source, target);
+                for (int i = 0; i < BUCKET_SIZE;)
+                {
+                    ref CastCacheEntry pEntry = ref Element(table, index);
+
+                    // must read in this order: version -> entry parts -> version
+                    // if version is odd or changes, the entry is inconsistent and thus ignored
+                    int version = Volatile.Read(ref pEntry._version);
+                    nuint entrySource = pEntry._source;
+
+                    // mask the lower version bit to make it even.
+                    // This way we can check if version is odd or changing in just one compare.
+                    version &= ~1;
+
+                    if (entrySource == source)
+                    {
+                        nuint entryTargetAndResult = Volatile.Read(ref pEntry._targetAndResult);
+                        // target never has its lower bit set.
+                        // a matching entryTargetAndResult would the have same bits, except for the lowest one, which is the result.
+                        entryTargetAndResult ^= target;
+                        if (entryTargetAndResult <= 1)
+                        {
+                            if (version != pEntry._version)
+                            {
+                                // oh, so close, the entry is in inconsistent state.
+                                // it is either changing or has changed while we were reading.
+                                // treat it as a miss.
+                                break;
+                            }
+
+                            return (CastResult)entryTargetAndResult;
+                        }
+                    }
+
+                    if (version == 0)
+                    {
+                        // the rest of the bucket is unclaimed, no point to search further
+                        break;
+                    }
+
+                    // quadratic reprobe
+                    i++;
+                    index = (index + i) & TableMask(table);
+                }
+            }
+            return CastResult.MaybeCast;
+        }
+
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        private static extern object IsInstanceOfAny_NoCacheLookup(void* toTypeHnd, object obj);
+
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        private static extern object ChkCastAny_NoCacheLookup(void* toTypeHnd, object obj);
+
+        // IsInstanceOf test used for unusual cases (naked type parameters, variant generic types)
+        // Unlike the IsInstanceOfInterface and IsInstanceOfClass functions,
+        // this test must deal with all kinds of type tests
+        [DebuggerHidden]
+        [StackTraceHidden]
+        [DebuggerStepThrough]
+        [MethodImpl(MethodImplOptions.AggressiveOptimization)]
```

Comment:
Ok, I see - it is because of this `[MethodImpl(MethodImplOptions.AggressiveOptimization)]`. I do not think we should have AggressiveOptimization on these.

Reply:
I think I mentioned this. Tiering eventually arrives to the same codegen as jitting, but there is some minor perf penalty. Probably because methods are located/aligned differently. I am not sure what else could be different. 
I am ok with removing AggressiveOptimization.

---
In reply to: [369332603](https://github.com/dotnet/runtime/pull/1068#discussion_r369332603) [](ancestors = 369332603)